### PR TITLE
L2-329: Manage errors stake neuron

### PR DESCRIPTION
**Pull request by lmuntanerDfinity**
# Motivation

Stake Neuron flow: manage edge cases, show errors gracefully and give a better UX on errors.

# Changes

* Add "invalidStates" array in the CreateNeuronModal. Error should be shown and modal closed on unexpected edge cases.
* Show busy screen when updating delay and staking neuron.
* Manage errors in services instead of component during the flow.
* Components ConfirmDissolveDelay, SetDissolveDelay and EditFollowNeurons are not rendered without a neuron.

### CODE REVIEW

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte [hunk 1]
```diff
@@ -68,40 +97,32 @@
...
-  <!-- TODO: Manage edge case: https://dfinity.atlassian.net/browse/L2-329 -->
-  {#if currentStepName === "StakeNeuron"}
-    <!-- we spare a spinner for the selectedAccount within StakeNeuron because we reach this step once the selectedAccount has been selected -->
-    {#if selectedAccount !== undefined}
-      <StakeNeuron
-        account={selectedAccount}
-        on:nnsNeuronCreated={goToDissolveDelay}
-      />
-    {/if}
+  {#if currentStepName === "StakeNeuron" && selectedAccount !== undefined}
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:103]: I made these changes on purpose - i.e. in my opinion the condition on the step of the wizard should only test the wizard name and not additional logic such as `&& selectedAccount !== undefined` or `newNeuron !== undefined`

In case of edge case, assuming the `currentStepName` is set and the additional clause is not set, the UI can become quirky and the user can be lost because no step and no feedback are displayed on screen.

In other words, for a smooth UI, a step should always be displayed.

That's why I deferred such logic as `selectedAccount !== undefined` or `newNeuron !== undefined` within their respective component and displayed for example spinner while data are loading.

How to you prevent these issues - i.e. how do you prevent no step to be rendered?

**Review by peterpeterparker — CHANGES_REQUESTED** [review]

File: frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte (hunk 1 above)
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:103]: I guess that it depends on what we want to do and the nature of the error.

At the moment, if we have the dissolve delay step and no neuron, it's not because we are fetching it, it's because something went wrong. Therefore, showing the spinner, will not be very helpful, because it will stay there forevever. We need to tell the user about it.

Another way is to also change the logic of when to move steps. Move steps earlier, before the previous step has finished. In this case, the spinner would make sense because in the end, the previous step would finish and we would have the neuron there. But we could show the spinner outside of the component, we show either the component or the spinner. But not the spinner inside the component.

I try to reduce the amount of complexity in components and adding the possiblity of `undefined` adds a case to be handled in the component that we could avoid. I guess this is a guideline we should agree on.

Also about when to move step. At the moment, we don't move until the step has finished totally. Therefore, the next step without neuron is an edge case where a Spinner is not information enough for the user.

Maybe a call with @mstrasinskis as well?

Comment (peterpeterparker) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:103]: - Spinner was only a placeholder, a skeleton or any other information can be used. 
- In my opinion and this particular case the complexity of testing if data are undefined or not should not be deferred to the parent because the parent is the wizard binder - i.e. beside nav within the wizard it should be kind of dumb

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte]: Discussed on Slack

**Comment by lmuntanerDfinity** [comment]
@peterpeterparker @mstrasinskis please review

File: frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte [hunk 2]
```diff
@@ -49,7 +51,34 @@
...
+    { stepName: "SetDissolveDelay", neuron: null },
+    { stepName: "ConfirmDissolveDelay", neuron: null },
+    { stepName: "EditFollowNeurons", neuron: null },
+  ];
+  $: {
+    newNeuron = $neuronsStore.find(({ neuronId }) => newNeuronId === neuronId);
+    const invalidState = invalidStates.find(({ stepName, neuron, account }) => {
+      return (
+        stepName === currentStep?.name &&
+        ((neuron === null && newNeuron === undefined) ||
```
Comment (peterpeterparker) [inline frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:71]: it's not "invalid" if neuron is null and newNeuron is not undefined and the contrary? likewise for account?

in short: it should not be`or`?

Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:71]: Maybe I should not deconstruct here. This is when the `invalidState.neuron === null && newNeuron === undefined`.

This is an invalid state:
```
  type InvalidState = {
    stepName: string;
    neuron?: null;
    account?: null;
  };
```

I could set the `newNeuron` initial value to `null` as well to make it comparable to the InvalidState.

Comment (peterpeterparker) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:71]: Not a problem with destructing, I probably just don't get it 🤪

```
const invalidStates: InvalidState[] = [
    { stepName: "StakeNeuron", account: null },
    ...
    ```
    
What does `account: null` here mean? Is that "the state is invalid if the stepName is StakeNeuron and if the selected account is null"?

If so, what about renaming `account` to something maybe more obvious like `invalidCondition`? Or maybe even a function?

```
type InvalidState = { stepName: string; isInvalid: <T>(obj: T) => boolean };
```

**Review by mstrasinskis — COMMENTED** [review]
Looks good! Just a couple of comments.

File: frontend/svelte/src/lib/components/neurons/SetDissolveDelay.svelte [hunk 3]
```diff
@@ -36,7 +36,7 @@
     dispatcher("nnsSkipDelay");
   };
   let neuronICP: bigint;
-  $: neuronICP = neuron?.fullNeuron?.cachedNeuronStake ?? BigInt(0);
+  $: neuronICP = neuron.fullNeuron?.cachedNeuronStake ?? BigInt(0);
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neurons/SetDissolveDelay.svelte:39]: good point!

File: frontend/svelte/src/lib/components/neurons/ConfirmDissolveDelay.svelte [hunk 4]
```diff
@@ -8,81 +8,68 @@
...
-      console.error(error);
-    } finally {
-      loading = false;
-    }
+    await updateDelay({
+      neuronId: neuron.neuronId,
+      dissolveDelayInSeconds: delayInSeconds,
+    });
+    stopBusy("update-delay");
+    dispatcher("nnsNext");
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neurons/ConfirmDissolveDelay.svelte:29]: good point, I thought the same, not sure the idea of throwing only in the services it matches always.

File: frontend/svelte/src/lib/components/neurons/ConfirmDissolveDelay.svelte [hunk 5]
```diff
@@ -8,81 +8,68 @@
...
+    <p>
+      {replacePlaceholders($i18n.neurons.icp_stake, {
+        $amount: formatICP(neuronICP),
+      })}
+    </p>
+  </div>
+  <div class="voting-power">
+    <h5>{$i18n.neurons.voting_power}</h5>
+    <p>
+      {votingPower({
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/components/neurons/ConfirmDissolveDelay.svelte:53]: yep!

File: frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte [hunk 6]
```diff
@@ -94,26 +134,29 @@
...
-  <!-- TODO: Manage edge case: https://dfinity.atlassian.net/browse/L2-329 -->
   {#if currentStep?.name === "SetDissolveDelay"}
-    <SetDissolveDelay
-      neuron={newNeuron}
-      on:nnsSkipDelay={goEditFollowers}
-      on:nnsConfirmDelay={goNext}
-      bind:delayInSeconds
-    />
+    {#if newNeuron !== undefined}
+      <SetDissolveDelay
```
[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:139]: yes, thanks, it got back in after rebasing and solving conflicts.

[the remark this answers is not in the document]
Comment (lmuntanerDfinity) [reply frontend/svelte/src/lib/modals/neurons/CreateNeuronModal.svelte:139]: done

**Review by mstrasinskis — APPROVED** [review]
Thanks 👍
